### PR TITLE
Add ACL example

### PR DESCRIPTION
**Pull request by satyaog**
Relates to #115

### CODE REVIEW

**Comment by satyaog** [comment]
https://mila-docs-staging--121.org.readthedocs.build/en/121/Userguide.html#sharing-data-with-acls

**Comment by satyaog** [comment]
@obilaniu , could you quickly look if this is the format you had in mind?

**Review by obilaniu — APPROVED** [review]
It's all good! Thanks for translating from Markdown!

File: docs/Userguide_sharing_data.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+* **Non**-recursive (!!!!)
+* May also grant ``:rx`` in unlikely event others listing your folders on the
+  path is not troublesome or desirable.
+
+----
+
+| In order to access a file, all folders from the root (``/``) down to the
+  parent folder in question must be searchable (``+x``) by the concerned user.
+  This is already the case for all users for folders such as ``/``, `/network`
```
Comment (btravouillon) [inline docs/Userguide_sharing_data.rst:52]:
```suggestion
  This is already the case for all users for folders such as ``/``, ``/network``
```

**Review by btravouillon — CHANGES_REQUESTED** [review]
One tiny typo.

Also, while this is very cool, this does not cover all topics of #115. I believe we won't be able to address all topics in a single PR, so I will remove the `resolves #115` to avoid case closure.

**Comment by btravouillon** [comment]
@vict0rsch: Any feedback on these ACL examples?

File: docs/Userguide_sharing_data.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
+Sharing Data with ACLs
+======================
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:2]: Maybe start by explaining what ACLs are, _and_ provide a link to learn (like https://www.thegeekdiary.com/unix-linux-access-control-lists-acls-basics/? Don't know if it's good, it's just the first Google result 😛)

File: docs/Userguide_sharing_data.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+As an illustrative example, to allow ``$USER`` to share with ``$USER2`` in
+``rwx`` a hierarchy ``/network/scratch/$USER/X/Y/Z/...`` with ACLs:
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:5]: I'm not sure "in `rwx`" would be comprehensible to a beginner

File: docs/Userguide_sharing_data.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+----
+
+| Grant **oneself** permissions to access any **future** files/folders created
+  by the other *(or oneself)*
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:10]: The `$USER` / `$USER2` paradigm in the intro is good to understand what's up, this `oneself` / `other` / `another` version is less clear IMHO

Comment (vict0rsch) [reply docs/Userguide_sharing_data.rst:10]: Also, I couldn't understand this sentence before the 4th read.

Is this saying: "how to make sure _I_ (`$USER`), will be able to `rwx` the files created by someone else (`$USER2`) if they put anything in `/network/scratch/$USER/X/Y/Z/`"

If that's the case, I think the first item in this tutorial should be how to _allow_ `$USER2` to even have access to `/network/scratch/$USER/X/Y/Z/`

File: docs/Userguide_sharing_data.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+| (``-d`` renders this permission a "default" / inheritable one)
+
+.. code-block:: bash
+
+   setfacl -Rdm user:$USER:rwx  /network/scratch/$USER/X/Y/Z/
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:15]: I'm guessing `-R` means recursive? `-d` is explained, what about `-m`?

File: docs/Userguide_sharing_data.rst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  by the other *(or oneself)*
+| (``-d`` renders this permission a "default" / inheritable one)
+
+.. code-block:: bash
+
+   setfacl -Rdm user:$USER2:rwx /network/scratch/$USER/X/Y/Z/
+
+----
+
+| Grant **another** permission to access any **existing** files/folders created
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:29]: I would put how to handle *existing files* _before_ *future files*

File: docs/Userguide_sharing_data.rst [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  their creation.
+
+.. code-block:: bash
+
+   setfacl -Rm  user:$USER2:rwx /network/scratch/$USER/X/Y/Z/
+
+----
+
+| Grant **another** permission to search through one's hierarchy down to the
+  shared location in question.
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:42]: Why would we want to do that? What does "search" mean? Only `ls`? What about reading?

File: docs/Userguide_sharing_data.rst [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+* **Non**-recursive (!!!!)
+* May also grant ``:rx`` in unlikely event others listing your folders on the
+  path is not troublesome or desirable.
+
+----
+
+| In order to access a file, all folders from the root (``/``) down to the
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:50]: Is this part of the previous section on "searching"? It _looks_ like it's its own different example, but the content seems like it's part of the previous block

**Comment by vict0rsch** [comment]
Thank you for working on this! I've made a bunch of comments. I think this is still a little raw and rough for beginners to understand completely though. 

In particular, I think this PR needs an explicit section on **Projects** and the ideal workflow for collaborators not to get confused with permissions and people not having access to others' data etc. I'm thinking about a "Good practices for Projects" section with the elements mentioned in #115: 

* group creation request to it-support
* profs approve
* provide usernames
* **how to setup ACLs for user groups**

(On that note, can you not make it default that any and all members of a group have `rwx` on the project's folder `/network/projects/my_project` so that by default people don't have to do anything with ACLs?)

File: docs/Userguide_sharing_data.rst [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+.. code-block:: bash
+
+   setfacl -m   user:$USER2:x   /network/scratch/$USER/X/Y/
+   setfacl -m   user:$USER2:x   /network/scratch/$USER/X/
+   setfacl -m   user:$USER2:x   /network/scratch/$USER/
+
+----
+
+| To bluntly allow **all** users to search through a folder (**think twice!**),
```
Comment (vict0rsch) [inline docs/Userguide_sharing_data.rst:65]: What if I want them to be able to copy (read) a file? Or write?

Comment (vict0rsch) [reply docs/Userguide_sharing_data.rst:65]: (_I_ personally know, I'm saying those examples should be there too 😄)

File: docs/Userguide_sharing_data.rst (hunk 4 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:10]: @vict0rsch There is a careful reasoning I made for executing those commands in that specific order, it's not by accident.

The idea behind granting *yourself* access to all future files created by (you or other) *first* is that if you don't do that, then in the gap between your execution of that command and the next one, that other person could race you to create files you mysteriously can't delete.

Granting rights to even access the shared workspace to the other person *last* ensures that all ACLs have been readied beforehand and no race condition is possible at the moment of the "unlock".

File: docs/Userguide_sharing_data.rst (hunk 5 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:15]: Yes, `-R` is recursive. `-m` means "modify" (add that ACL rule to the set), as opposed to "set" (replace all ACL rules with the ones provided).

More on that in `man setfacl`. I think people should develop at least some reflex to consult the man-pages, it's almost always `man commandname`.

File: docs/Userguide_sharing_data.rst (hunk 4 above)
Comment (vict0rsch) [reply docs/Userguide_sharing_data.rst:10]: So this is for security reasons? It's interesting. Put this^ as a `..note:` no? A good doc is also a good place to learn 😛 But I understand now thanks

File: docs/Userguide_sharing_data.rst (hunk 6 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:29]: See a comment above

File: docs/Userguide_sharing_data.rst (hunk 7 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:42]: To put it in extremely plain language, Unix separates the concept of "traversing" a directory from "reading" it.

To execute `ls /network/scratch/$USER/a/b/c/workspace/`, a `$USER2` **does not** require "read" (`r`) permission on _anything_ except `workspace/` itself. A `$USER2` **does** however require "search" permissions on **all** directories from the root down to `workspace/`. Unix overloads the "executable" (`x`) permission (which makes no sense on directories) for "search".

The rules are somewhat baroque but are elaborated upon in greater detail in `man path_resolution`. The summary that's important for most people is what I wrote above, and what my proposed `setfacl` commands arrange for.

**Comment by btravouillon** [comment]
@obilaniu would you have some time available to update the PR in regards to @vict0rsch comments?

File: docs/Userguide_sharing_data.rst [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  This is already the case for all users for folders such as ``/``, ``/network``
+  and ``/network/scratch``, but users must explicitly grant access to some or
+  all users by adding ACLs for ``/network/scratch/$USER``, ``$HOME`` and
+  subfolders.
+
+.. code-block:: bash
+
+   setfacl -m   user:$USER2:x   /network/scratch/$USER/X/Y/
+   setfacl -m   user:$USER2:x   /network/scratch/$USER/X/
+   setfacl -m   user:$USER2:x   /network/scratch/$USER/
```
Comment (btravouillon) [inline docs/Userguide_sharing_data.rst:61]: Just noticed the path should be `/network/scratch/$U{USER:0:1}/$USER/`

Comment (liellnima) [reply docs/Userguide_sharing_data.rst:61]: Hey, this example here is very helpful - immediately solved my questions and it became quickly clear that one has to make the subdirectories accessible/searchable first!

**Comment by btravouillon** [comment]
@obilaniu would you have some time available to update the PR in regards to @vict0rsch comments?

File: docs/Userguide_sharing_data.rst (hunk 9 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:65]: We will make ourselves a task for this in #128.

File: docs/Userguide_sharing_data.rst (hunk 3 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:5]: We will make ourselves a task for this in #128.

**Review by obilaniu — APPROVED** [review]

**Comment by vict0rsch** [comment]
@saleml does not know what ACLs are. He'll be a great tester for this PR :)

File: docs/Userguide_sharing_data.rst (hunk 4 above)
Comment (btravouillon) [reply docs/Userguide_sharing_data.rst:29]: It's not for security reasons but for safety reasons. Not sure we should add too much explanations about the examples.

File: docs/Userguide_sharing_data.rst (hunk 5 above)
Comment (btravouillon) [reply docs/Userguide_sharing_data.rst]: People need to use the man-pages to get further explanation about commands and parameters. The documentation emphasizes the important parameters in the context, but we should not have to duplicate the man pages by explaining each parameter.

File: docs/Userguide_sharing_data.rst (hunk 4 above)
Comment (obilaniu) [reply docs/Userguide_sharing_data.rst:29]: @btravouillon In the end I did supply such an explanation, see [here](https://mila-docs-staging--121.org.readthedocs.build/en/121/Userguide.html#sharing-data-with-acls)

I think the current notes strike the right balance between explaining why this should probably not be skipped and delving too deep in the weeds.

Comment (btravouillon) [reply docs/Userguide_sharing_data.rst:29]: Well, the comment was not resolved. I just noticed while doing the (hopefully) last review.

**Review by btravouillon — APPROVED** [review]
Looks good to me. 👍

**Pull request merged**